kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/BatchAwareMemoryRecordsBuilder.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.records;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.record.CompressionType;
+import org.apache.kafka.common.record.EndTransactionMarker;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.record.SimpleRecord;
+import org.apache.kafka.common.record.TimestampType;
+import org.apache.kafka.common.utils.ByteBufferOutputStream;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A builder of {@link MemoryRecords} that can cope with multiple batches.
+ * (Apache Kafka's own {@link MemoryRecordsBuilder} assumes a single batch).
+ */
+public class BatchAwareMemoryRecordsBuilder {
```

Comment:
consider implementing AutoCloseable like MRB

Reply:
I'm on the fence about this one. It's not clear to me what value it really adds for people using this class. It adds confusion (should I call `close()` or `build()`?). Looking at what `MRB.close()` does is closes the compressing streams, so I guess it could free some native resources. So probably we ought to `close()` things nicely. But does that mean we need to ensure that the `Collector`s which use BAMRBs call `close()` when the Stream is `close()`d? I guess it does... in which case: Good catch!

Reply:
I went down a deep rabbit hole on this one, and have ended up adding the ability to close the BAMRB to the `Collectors`, as well as being more careful with the Streams we return. This means we should be OK even in the event of an exception while terminating the stream, so long as the user of the Stream does the right thing with `try/finally`.